app/frontend/src/service/client.ts

```diff
@@ -8,9 +9,34 @@ import { SSOPromiseClient } from "../pb/sso_grpc_web_pb";
 
 const URL = process.env.REACT_APP_API_BASE_URL;
 
+let _unauthenticatedErrorHandler: (e: object) => void = () => {};
+export const setUnauthenticatedErrorHandler = (f: (e: object) => void) => {
+  _unauthenticatedErrorHandler = f;
+};
```

Comment:
@darrenvong your favourite 😈 

But seriously, I can't think of a better way since interceptors are outside the tree and we're using context now.

Reply:
After more thoughts, I think I prefer handling errors within the each query route more as this file is essentially a black box that we can't test easily since we've been mocking it out completely, so it'd be nice to avoid putting any logic here.

Reply:
That's true, but I a bit hesitant to put `if unauthenticated, log out` every time we use the API. It's code duplication and prone to forgetting.

Maybe we should use mobx? Or would that still have the same issue - we would be able to call functions on the store outside the react tree, but it would still be logic in this file.

Reply:
Yeah MobX would solve the memory issue but not the testability issue.

I think creating an [auth error boundary](https://reactjs.org/blog/2017/07/26/error-handling-in-react-16.html), have the logout logic there and use it to wrap the rest of the app might be the best way to go. This might be the rare case where we'll have to use a class component as error boundaries component still doesn't have an equivalent in function component AFAIK!

Reply:
Speaking of which, this would be a nice place to put some generic "sorry, we encountered some issues" UI rather than letting the app crash for errors that we might not have seen before and so don't have a specific strategy to handle it, so I think this is a good idea.

Reply:
Won't that still require logic in every hook? Because at the moment, the error is caught and displayed in an alert, so it won't make it to the error boundary.

Reply:
Isn't it only caught because of this interceptor? I was thinking if you don't catch it here, React Query doesn't have error handling if the service promise throws

Reply:
If you mean the existing the Redux code then yeah, will probably have to remove the catches and let it propagate to the error boundary, but should be ok if we are eventually dismantling it anyway?

Reply:
You know what, I will try removing all the alerts and just put one above the router and catch there. Individual components can then still catch if they need to for a particular reason but in general they don't I think

Reply:
After experimenting some more, still haven't come up with a solution. Error boundaries don't catch errors in async logic. You can [use a fake 'setState' inside `catch`](https://medium.com/trabe/catching-asynchronous-errors-in-react-using-error-boundaries-5e8a5fd7b971), but you have to put that in every component. It would apparently work in concurrent mode, but that is still experimental.

I agree it's bad to have this mystery bit that is mocked out and will not be tested. Maybe the question is, do we prefer 10 or so lines of untested code or putting `authActions.checkForAuthError(error)` in every hook? I keep going back and forth thinking about it...

Reply:
I didn't know that but makes sense when explained in that post, so that's good to know.

And yeah, I'm torn about this too, although I'm leaning more towards repetition (it's one extra line if extracted into a function) so we can test this and have increased confidence that things work as intended.

Reply:
I think this also brings into question our testing strategy - the fact that we completely mock this out means that if we do have a legit use case where things would only work with interceptors, we can't use them with confidence due to the lack of tests.

Maybe we work harder on only mocking the `client.<grpcMethod>()` would be a better answer, because that way, we can still check interceptors are applied correctly and the clients are called with the right params etc. They seem like an elegant tool we lose/can't use confidently because we're "slacking" on our tests.

Reply:
What about just testing `AuthInterceptor.intercept`? We can mock the `invoker` and pass it in. We could also mock `setUnauthenticatedErrorHandler` and test the useEffect in `useAuthStore`. So, testing the two halves seperately. An integration test of them together would be very difficult, I think.

Reply:
I don't think it'd be too difficult (although probably awkward since we'll have to work with those `*Res` classes...) since the grpc methods on the client returns back an object rather than a buffer response.

Otherwise, the `setUnauthenticatedErrorHandler` mock is a bit better since it tests the useEffect sets the right thing and we'll have to assume the blackbox works from that point on. Not sure I'd bother with testing the `AuthInterceptor` if you mean calling `intercept` directly with a mocked `invoker` succeeding/throwing etc, since everything is so instrumented at that point and it's not how the code uses it, I don't see much value in maintaining extra tests for that.

Reply:
I'm not sure it would work (or I'm not understanding). The clients are created with `new XPromiseClient(opts)` which is library code and `opts` contains the interceptor which runs mysteriously in the library. I don't see what you could mock to make it work easily.

Reply:
If we let it run as usual, unless I overlooked something, the only reason it'd fail would be because it's trying to make network calls right? So if we can mock enough to stop it trying to hit a real API, the rest of the code should run fine?

Reply:
Yes, but that would involve mocking grpc-web internals. Not sure where to start, even looking at the source!

Reply:
For now I did the `setUnauthenticatedErrorHandler` mock test

Reply:
I'd have thought mocking the grpc methods on the client's prototype would be enough, as I assume that's where it starts making requests?

Reply:
You can do that, but it won't use the interceptor. [That happens here!](https://github.com/grpc/grpc-web/blob/e88fdc809d9ccbf1c9dae701ee14cce53be38b9e/javascript/net/grpc/web/grpcwebclientbase.js#L103)

Reply:
Wow, that's annoying, the whole library is such a pile of 💩  I guess we are screwed with the blackbox then 😢

Reply:
Actually after looking at their implementation, I changed my mind about what I said on unit testing the interceptor, since it doesn't look like they're doing anything special other than calling them all in a loop.

So yeah, let's do that and I'm happy to wrap this up!